### ISSUE TITLE
Dynamic Buffer for entity

### ISSUE THREAD

**Issue by RomanZhu** [SanderMertens/flecs#8]
Hey :D

When working with Unity ECS I found Dynamic buffers to be really usefull:
https://docs.unity3d.com/Packages/com.unity.entities@0.0/manual/dynamic_buffers.html

What do you think about it?
For example to handle collisions, there can be multiple collision hits at same frame. And to not put gameplay logic into collision detection we would separate dealing damage based on those hits into another system. And a bunch more usecases.

**Comment by jesta88** [comment]
This is also a requirement for us to re-simulate a buffer of inputs in a networking scenario.

**Comment by SanderMertens** [comment]
You can do this in flecs with the `EcsArray` type, but since this is C and not C#, you will have to do a little bit of memory management yourself ;)

The `EcsArray` is like the `DynamicBuffer` in that it automatically grows if you add elements to it. It is the data type that flecs uses internally to store the components. This is the header of the type: https://github.com/SanderMertens/flecs/blob/master/include/util/array.h

A small code example:
```
EcsArrayParams params = {.element_size = sizeof(MyType)};
EcsArray *arr = ecs_array_new(&params, 2); // create array with 2 initial elements

// Add a few elements
MyType *elem1 = ecs_array_add(&arr, &params);
MyType *elem2 = ecs_array_add(&arr, &params);
MyType *elem3 = ecs_array_add(&arr, &params); // third element, buffer will grow
```
A few things you'll note:
First, `EcsArrayParams` is a type that stores static data of the array. The reason that this is not part of the `EcsArray` object is to conserve memory. If you have many `EcsArray` objects with the same element type, you only have to store the `EcsArrayParams` once. Typically, you have a number of global const variables that define the `EcsArrayParams` for your different array types.

Secondly, note that the `ecs_array_add` accepts a _pointer to the array_. This is because as the array grows, it may be reallocated, and the address may change.

Now, when you use `EcsArray` in a component, you need to make sure that:
1. you initialize it when the component is created
2. you free it when the component is deleted

For this, you can use `EcsOnAdd` and `EcsOnRemove` systems. For example, if I have this component:

```c
struct MyComponent {
    EcsArray *my_array;
};
```

I can create two systems like this:
```c
ECS_COMPONENT(world, MyComponent);
ECS_SYSTEM(world, InitMyComponent, EcsOnAdd, MyComponent);
ECS_SYSTEM(world, DeinitMyComponent, EcsOnRemove, MyComponent);
```

That I can implement like this:
```c
EcsArrayParams params = {.element_size = sizeof(MyComponent)};

void InitMyComponent(EcsRows *rows) {
    MyComponent *data = ecs_column(rows, MyComponent, 1);
    for (int i = rows->begin; i < rows->end; i ++) {
        data[i].my_array = ecs_array_new(&my_component_params, 0);
    }
}

void DeinitMyComponent(EcsRows *rows) {
    MyComponent *data = ecs_column(rows, MyComponent, 1);
    for (int i = rows->begin; i < rows->end; i ++) {
        ecs_array_free(data[i].my_array);
    }
}
```

It's a bit of work, but it gives you maximum control over memory and a very low footprint. Future features (like reflection) could automate this, but for now this is probably the best way to do it.

**Comment by jesta88** [comment]
Sounds good! Being explicit with these things is a plus in my opinion. That's why Unity is deprecating all injection/reflection from ECS.

**Comment by RomanZhu** [comment]
Yep, looks great.

OnAdd and OnRemove are callled immediately when component is added/removed or in order of systems?

**Comment by SanderMertens** [comment]
They are called immediately, just like `EcsOnSet`, which is called immediately when you can `ecs_set`.

**Comment by teksoc-1** [comment]
Hi, I have been reviewing the flecs framework ( and really like it thus far ) for some time now and when doing some test cases of deleting entities and having OnRemove systems registered with components of the deleted entities, the entities seem to not get deleted, and will see them again next iteration. If I do not register OnRemove systems, the entities get deleted as expected, however, when the OnRemove Systems are applied ( and respectively called as expected ), the entities do not get deleted. Is there something that could perhaps be wrong with my set up?

**Comment by teksoc-1** [comment]
Also in the context of deletion of entities, I was using the profiler tool in Xcode, to see memory behavior when creating N entities and deleting the same N entities in while ecs_progress() loop, I observed memory consumption continue to grow and what appeared to be a leak? A little further digging, it seemed that perhaps the staged columns for the data stage where not being freed; I applied a small patch of code in merge_commits() to iterate over the staged columns of the data stage and apply eco_vector_free() do the .data of each staged column for the merged entities. While that seemed to fix the perceived memory leak ( at least to the extent that it gets a nice happy green check mark under Xcode instruments, and I don't visually see the memory continue to grow ), I have no idea if that is even correct, or any unintended side effects I'm not considering.
